core/src/main/scala/com/evolutiongaming/catshelper/ReadWriteRef.scala

```diff
@@ -0,0 +1,193 @@
+   */
+  def read: Resource[F, A]
+
+  /**
+   * "Write" operation. Requires exclusive access.
+   * `write.use(…)` blocks semantically while [[read]] or another [[write]] is in use.
+   */
+  def write: Resource[F, ReadWriteRef.Upd[F, A]]
+}
+
+object ReadWriteRef {
+  /**
+   * A "resource" of [[ReadWriteRef.write `write`]].
+   * `(A => F[A])` part lets you read current value and produce a new one, effectfully.
+   * `F[A]` in the return part lets you use new `A` in the outer context, e.g. return it from `use`.
+   */
+  trait Upd[F[_], A] extends ((A => F[A]) => F[A])
+
+  object Upd {
+    implicit final class SelfOps[F[_], A](val self: Upd[F, A]) extends AnyVal {
+      /** A shorthand for non-effectful updates. */
+      def plain(f: A => A)(implicit F: Applicative[F]): F[A] = self(a => F.pure(f(a)))
+
+      /** A shorthand for non-effectful update that ignores previous value. */
+      def set(a: A)(implicit F: Applicative[F]): F[A] = self(_ => F.pure(a))
+    }
+  }
+
+  final class PartialApply[F[_]: Concurrent] {
+    def of[A](init: A): F[ReadWriteRef[F, A]] = ReadWriteRef.of[F, A](init)
+  }
+
+  def apply[F[_]: Concurrent]: PartialApply[F] = new PartialApply[F]
+
+  def of[F[_], A](init: A)(implicit F: Concurrent[F]): F[ReadWriteRef[F, A]] = {
+    sealed trait Pending
+    case class PendingRead(listener: A => F[Unit]) extends Pending
+    case class PendingWrite(listener: A => F[Unit]) extends Pending
+
+    case class State(
+      a: A,
+      inUse: Int = 0, // -1 -- exclusive write, 0 -- not in use, >0 -- readers
+      pending: Queue[Pending] = Queue.empty,
+    ) {
+      def canRead: Boolean = inUse >= 0 && pending.isEmpty
+      def canWrite: Boolean = inUse == 0 && pending.isEmpty
+
+      def without(p: Pending): State = {
+        // Might want to optimise this later.
+        copy(pending = pending.filterNot(_ == p))
+      }
+    }
+
+    Ref[F].of(State(init)) map { stateRef =>
+      // The `release` function "commits" A back to the state, adjusts `inUse` count,
+      // and notifies pending listeners.
+      //
+      // IMPORTANT:
+      //   This function must be called only from uncancellable blocks (e.g. Resource release).
+      def release(a: A, useChange: Int => Int): F[Unit] = stateRef
+        .modify { s0 =>
+          val inUse = useChange(s0.inUse)
+
+          s0.pending match {
+            case PendingWrite(listener) +: rest if inUse == 0 =>
+              val s1 = s0.copy(a = a, inUse = -1, pending = rest)
+              val effect = listener(a).start.void
+              s1 -> effect
+
+            case _ =>
+              // A slightly optimized way to split the pending ops in two and start the first batch
+              @tailrec def startPending(effect: F[Unit], n: Int, ps: Queue[Pending]): (F[Unit], Int, Queue[Pending]) = {
+                ps match {
+                  case PendingRead(listener) +: tail => startPending(effect <* listener(a).start, n + 1, tail)
```

Comment:
does that `match` work with `dequeue` underneath ?

Reply:
It should, but let me check it.

Reply:
Hah! It does `(head, tail)` and while `tail` is quite efficient, `head` may be sub-optimal in some cases. Let me improve that.

Reply:
Done in e87a7bc.